### PR TITLE
Optimise runtime lock performance

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

*Optimise runtime lock performance*

* Fix lock code when a timeout occurs. In this case it was possible that a queued writer was not notified.
* Disable multi-insert optimization in AQL when it is executed as part of a streaming transaction. This is necessary because the multi-insert optimization will start its own sub-transaction, and does not know about if it is invoked from an outer streaming transaction. If it is invoked from inside a streaming transaction that holds write locks and then starts its own transaction that tries to acquire the same locks, it will deadlock.
* (unimportant) Adjust a test so that it does not acquire a lock recursively on the same collection 3 times.

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [X] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [x] Backport for 3.11: *https://github.com/arangodb/arangodb/pull/19346*
  - [x] Backport for 3.10: *https://github.com/arangodb/arangodb/pull/19345*
  - [ ] Backport for 3.9: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 1]
```diff
@@ -1661,63 +1661,38 @@ Result RocksDBMetaCollection::applyUpdatesForTransaction(
...
+    TRI_ASSERT(false);
+    return TRI_ERROR_INTERNAL;
+  }
 
-      startTime = now;
-      waitTime = 1;
-    } else {
-      TRI_ASSERT(startTime > 0.0);
+  if (gotLock) {
+    // keep the lock and exit the loop
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:1683]: We don't have a loop anymore
```suggestion
    // keep the lock and exit
```

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 2]
```diff
@@ -1661,63 +1661,38 @@ Result RocksDBMetaCollection::applyUpdatesForTransaction(
...
+    return TRI_ERROR_NO_ERROR;
+  }
 
-      if (now > startTime + timeout) {
-        LOG_TOPIC("d1e52", TRACE, arangodb::Logger::ENGINES)
-            << "timed out after " << timeout << " s waiting for "
-            << AccessMode::typeString(mode) << " lock on collection '"
-            << _logicalCollection.name() << "'";
+  TRI_ASSERT(startTime > 0.0);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:1687]: `startTime` will always be 0.0 here, so the assertion will be violated.
We can as well remove the `startTime` variable entirely, as it does not serve any purpose anymore.

File: arangod/RocksDBEngine/RocksDBMetaCollection.h [hunk 3]
```diff
@@ -192,7 +191,7 @@ class RocksDBMetaCollection : public PhysicalCollection {
  protected:
   RocksDBMetadata _meta;  /// collection metadata
   /// @brief collection lock used for write access
-  mutable basics::ReadWriteLock _exclusiveLock;
+  mutable std::shared_timed_mutex _exclusiveLock;
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBMetaCollection.h:194]: If we are using `shared_timed_mutex` now, we must also include `shared_mutex` in the header.

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp (hunk 2 above)
Comment (kvahed) [reply arangod/RocksDBEngine/RocksDBMetaCollection.cpp:1687]: agree. removing

File: arangod/RocksDBEngine/RocksDBMetaCollection.h (hunk 3 above)
Comment (kvahed) [reply arangod/RocksDBEngine/RocksDBMetaCollection.h:194]: sure.

File: CHANGELOG [hunk 4]
```diff
@@ -1,7 +1,9 @@
 devel
 -----
 
-* Attempt to avoid busy looping when locking collections with a timeout
+* Optimise runtime performance of exclusive locks.
```
Comment (jsteemann) [inline CHANGELOG:4]:
```suggestion
* Optimize runtime performance of exclusive locks.
```

**Review by jsteemann — APPROVED** [review]
LGTM

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 5]
```diff
@@ -1661,63 +1659,35 @@ Result RocksDBMetaCollection::applyUpdatesForTransaction(
...
+  bool gotLock = false;
+  if (mode == AccessMode::Type::WRITE) {
+    gotLock =
+        _exclusiveLock.try_lock_for(std::chrono::duration<double>(timeout));
+  } else if (mode == AccessMode::Type::READ) {
+    gotLock = _exclusiveLock.try_lock_shared_for(
+        std::chrono::duration<double>(timeout));
+  } else {
+    // we should never get here
+    TRI_ASSERT(false);
```
Comment (MBkkt) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:1676]: Same assert as in the 1665 line. Could you remove one of them?

File: CHANGELOG [hunk 6]
```diff
@@ -37,6 +39,7 @@ devel
 * BTS-1490: query can use a lot more memory when using COLLECT WITH COUNT.
 
 * Attempt to avoid busy looping when locking collections with a timeout
+>>>>>>> 38f46a7e32a48ba83467c529f5bab2ae0350db65
```
Comment (mpoeter) [inline CHANGELOG:42]:
```suggestion
>>>>>>> 38f46a7e32a48ba83467c529f5bab2ae0350db65
```

File: arangod/RocksDBEngine/RocksDBMetaCollection.cpp [hunk 7]
```diff
@@ -1661,63 +1659,37 @@ Result RocksDBMetaCollection::applyUpdatesForTransaction(
...
 /// @brief lock a collection, with a timeout
 ErrorCode RocksDBMetaCollection::doLock(double timeout, AccessMode::Type mode) {
-  uint64_t waitTime = 0;  // indicates that time is uninitialized
-  double startTime = 0.0;
-
   // user read operations don't require any lock in RocksDB, so we won't get
   // here. user write operations will acquire the R/W lock in read mode, and
   // user exclusive operations will acquire the R/W lock in write mode.
   TRI_ASSERT(mode == AccessMode::Type::READ || mode == AccessMode::Type::WRITE);
+  auto tous = std::chrono::duration_cast<std::chrono::microseconds>(
```
Comment (mpoeter) [inline arangod/RocksDBEngine/RocksDBMetaCollection.cpp:1666]: Can we please rename this to `timeout_us` or just `timeout`? I know the first one does not match our styleguide, but `tous` sounds really nonsensical to me (is this french?).

**Comment by mpoeter** [comment]
@MBkkt Yes, we need a RWLock that favors writers over readers. That's why we use our own implementation.

File: CHANGELOG (hunk 6 above)
Comment (kvahed) [reply CHANGELOG:42]: already fixed

File: lib/Basics/ReadWriteLock.cpp [hunk 8]
```diff
@@ -130,6 +130,20 @@ void ReadWriteLock::lockRead() {
...
+  }
+  std::unique_lock<std::mutex> guard(_reader_mutex);
+  while (true) {
+    if (tryLockRead()) {
+      return true;
+    }
+    _readers_bell.wait_until(guard, end);
+  }
+  return false;
+}
```
Comment (MBkkt) [inline lib/Basics/ReadWriteLock.cpp:146]: Is it expected that return false is impossible?

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: CHANGELOG (hunk 6 above)
Comment (mpoeter) [reply CHANGELOG:42]: Cannot confirm

**Review by mpoeter — CHANGES_REQUESTED** [review]

File: lib/Basics/ReadWriteLock.cpp (hunk 8 above)
Comment (kvahed) [reply lib/Basics/ReadWriteLock.cpp:146]: Yes

Comment (MBkkt) [reply lib/Basics/ReadWriteLock.cpp:146]: It's a bug.

Comment (MBkkt) [reply lib/Basics/ReadWriteLock.cpp:146]: It will be busy loop if timeout will be reached

Comment (mpoeter) [reply lib/Basics/ReadWriteLock.cpp:146]: Valery is right. The `while (true)` does not honor the timeout. You have to use a similar approach as in `tryLockWriteFor` with the `cv_status`

**Review by mpoeter — CHANGES_REQUESTED** [review]

File: lib/Basics/ReadWriteLock.cpp [hunk 9]
```diff
@@ -130,6 +129,22 @@ void ReadWriteLock::lockRead() {
   }
 }
 
+bool ReadWriteLock::tryLockReadFor(std::chrono::microseconds timeout) {
+  auto end_time = std::chrono::steady_clock::now() + timeout;
+  if (tryLockRead()) {
+    return true;
+  }
```
Comment (jsteemann) [inline lib/Basics/ReadWriteLock.cpp:136]: We can avoid calculating `end_time` if we reverse the order of these operations:
```suggestion
  if (tryLockRead()) {
    return true;
  }
  auto end_time = std::chrono::steady_clock::now() + timeout;  
```

File: lib/Basics/ReadWriteLock.cpp [hunk 10]
```diff
@@ -130,6 +129,22 @@ void ReadWriteLock::lockRead() {
...
+  std::unique_lock<std::mutex> guard(_reader_mutex);
+  std::cv_status status(std::cv_status::no_timeout);
+  while (std::cv_status::no_timeout == status) {
+    if (tryLockRead()) {
+      return true;
+    }
+    status = _readers_bell.wait_until(guard, end_time);
+  }
```
Comment (jsteemann) [inline lib/Basics/ReadWriteLock.cpp:144]:
```suggestion
  std::cv_status status(std::cv_status::no_timeout);
  std::unique_lock<std::mutex> guard(_reader_mutex);
  do {
    if (tryLockRead()) {
      return true;
    }
    status = _readers_bell.wait_until(guard, end_time);
  } while (std::cv_status::no_timeout == status);
```

**Review by mpoeter — APPROVED** [review]

**Review by MBkkt — APPROVED** [review]

**Review by kvahed — COMMENTED** [review]
LGTM

**Pull request merged**